openstack/networking/v2/extensions/bgpvpns/requests.go

```diff
@@ -0,0 +1,472 @@
+	if err != nil {
+		r.Err = err
+		return
+	}
+	resp, err := client.Post(ctx, createNetworkAssociationURL(client, id), b, &r.Body, nil)
+	_, r.Header, r.Err = gophercloud.ParseResponse(resp, err)
+	return
+}
+
+// GetNetworkAssociation retrieves a specific network association by BGP VPN id
+// and network association id.
+func GetNetworkAssociation(ctx context.Context, c *gophercloud.ServiceClient, bgpVpnID string, id string) (r GetNetworkAssociationResult) {
+	resp, err := c.Get(ctx, getNetworkAssociationURL(c, bgpVpnID, id), &r.Body, nil)
+	_, r.Header, r.Err = gophercloud.ParseResponse(resp, err)
+	return
+}
+
+// DeleteNetworkAssociation deletes a specific network association by BGP VPN id
+// and network association id.
+func DeleteNetworkAssociation(ctx context.Context, c *gophercloud.ServiceClient, bgpVpnID string, id string) (r DeleteNetworkAssociationResult) {
+	resp, err := c.Delete(ctx, deleteNetworkAssociationURL(c, bgpVpnID, id), nil)
+	_, r.Header, r.Err = gophercloud.ParseResponse(resp, err)
+	return
+}
+
+// ListRouterAssociationsOptsBuilder allows extensions to add additional
+// parameters to the ListRouterAssociations request.
+type ListRouterAssociationsOptsBuilder interface {
+	ToRouterAssociationsListQuery() (string, error)
+}
+
+// ListRouterAssociationsOpts allows the filtering and sorting of paginated
+// collections through the API.
+type ListRouterAssociationsOpts struct {
+	Fields []string `q:"fields"`
+	Limit  int      `q:"limit"`
+	Marker string   `q:"marker"`
+}
+
+// ToRouterAssociationsListQuery formats a ListRouterAssociationsOpts into a
+// query string.
+func (opts ListRouterAssociationsOpts) ToRouterAssociationsListQuery() (string, error) {
+	q, err := gophercloud.BuildQueryString(opts)
+	if err != nil {
+		return "", err
+	}
+	return q.String(), nil
+}
+
+// ListRouterAssociations pages over the router associations of a specified
+// BGP VPN.
+func ListRouterAssociations(c *gophercloud.ServiceClient, id string, opts ListRouterAssociationsOptsBuilder) pagination.Pager {
+	url := listRouterAssociationsURL(c, id)
+	query, err := opts.ToRouterAssociationsListQuery()
+	if err != nil {
+		return pagination.Pager{Err: err}
+	}
+	url += query
+	return pagination.NewPager(c, url, func(r pagination.PageResult) pagination.Page {
+		p := RouterAssociationPage{pagination.MarkerPageBase{PageResult: r}}
+		p.MarkerPageBase.Owner = p
+		return p
+	})
+}
+
+// CreateRouterAssociationOptsBuilder allows extensions to add additional
+// parameters to the CreateRouterAssociation request.
+type CreateRouterAssociationOptsBuilder interface {
+	ToRouterAssociationCreateMap() (map[string]interface{}, error)
+}
+
+// CreateRouterAssociationOpts represents options used to create a BGP VPN
+// router association.
+type CreateRouterAssociationOpts struct {
+	RouterID             string `json:"router_id" required:"true"`
+	AdvertiseExtraRoutes *bool  `json:"advertise_extra_routes,omitempty"`
```

Comment:
Funny, [this field is undocumented](https://docs.openstack.org/api-ref/network/v2/#id857).

Reply:
It's documented in PUT method. I've just checked whether `project_id` can be set as well for cases, when admins need to create a resource. I'll add this option as well. Otherwise cloud admins will get `Request forbidden: [POST https://network-3.qa-de-1.cloud.sap/v2.0/bgpvpn/bgpvpns/920b4f41-6553-4ec5-ad87-f3bf156a7650/router_associations], error message: {"NeutronError": {"type": "NotAuthorized", "message": "Not authorized.", "detail": ""}}`